man/man3/pmda.3
'\"macro stdmacro
.\"
.\" Copyright (c) 2000-2004 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.\" add in the -me strings for super and subscripts
.ie n \{\
.       ds [ \u\x'-0.25v'
.       ds ] \d
.       ds { \d\x'0.25v'
.       ds } \u
.\}
.el \{\
.       ds [ \v'-0.4m'\x'-0.2m'\s-3
.       ds ] \s0\v'0.4m'
.       ds { \v'0.4m'\x'0.2m'\s-3
.       ds } \s0\v'-0.4m'
.\}
.TH PMDA 3 "PCP" "Performance Co-Pilot"
.SH NAME
\f3PMDA\f1 \- introduction to the Performance Metrics Domain Agent support library
.SH "C SYNOPSIS"
.ft 3
.ad l
.hy 0
#include <pcp/pmapi.h>
.br
#include <pcp/pmda.h>
.sp
.ft 1
\& ... assorted routines ...
.ft 3
.sp
cc ... \-lpcp_pmda \-lpcp
.hy
.ad
.ft 1
.SH DESCRIPTION
.de CR
.ie t \f(CR\\$1\f1\\$2
.el \fI\\$1\f1\\$2
..
To assist in the development of Performance Metric Domain Agents
.RB ( PMDA s)
for the Performance Co-Pilot (PCP),
a procedural interface is provided that extends the Performance Metrics
Application Programming Interface (\c
.BR PMAPI (3))
library.  These procedures are designed to enable a programmer to quickly
build a
PMDA
which can then be tested and refined.  However, this also
implies that a
PMDA
has a particular structure which may not be suitable for
all PMDA implementations.
.PP
Once you are familiar with the PCP and
PMDA
frameworks, you can quickly implement a new
PMDA
with only a few data structures and functions.  This is covered in far greater
detail in the
.IR "Performance Co-Pilot Programmer's Guide" .
.PP
A
PMDA
is responsible for a set of performance metrics, in the sense that it must
respond to requests from
.BR pmcd (1)
for information about performance metrics, instance domains, and instantiated
values.
.PP
This man page contains sections of the
.B simple PMDA
which is located at
.IR $PCP_PMDAS_DIR /simple.
.SH COMMUNICATING WITH PMCD
Two approaches may be used for connecting a
PMDA
to a
.BR pmcd (1)
process.  A Dynamic Shared Object (DSO) can be attached by
.BR pmcd (1)
using
.BR dlopen (3)
when the
.BR pmcd (1)
process is started.  A procedural interface referenced through a shared data
structure is used to handle requests from
.BR pmcd (1)
to the PMDA .
.PP
The preferred approach is for a separate process (daemon) to communicate with
.BR pmcd (1)
using the Performance Data Units (PDU) Inter-Process Communication (IPC)
protocol.
.PP
All PMDAs
are launched and controlled by the
.BR pmcd (1)
process on the local host.  The requests from the clients are received by
.BR pmcd (1)
and forwarded to the appropriate PMDAs.
Responses, when required, are returned through
.BR pmcd (1)
to the clients.  The requests (PDUs) that may be sent to a
PMDA
from
.BR pmcd (1)
are
.BR PDU_FETCH ,
.BR PDU_PROFILE ,
.BR PDU_INSTANCE_REQ ,
.BR PDU_DESC_REQ ,
.BR PDU_TEXT_REQ
and
.BR PDU_RESULT .
If the PMDA implements any dynamic metrics it may also receive
.BR PDU_PMNS_CHILD ,
.BR PDU_PMNS_IDS ,
.B PDU_PMNS_NAMES
and
.B PDU_PMNS_TRAVERSE
PDUs.
.SH DEFAULT CALLBACKS FOR HANDLING PDUs
To allow a consistent framework,
.BR pmdaMain (3)
can be used by a daemon
PMDA
to handle the communication protocol using the same callbacks as a DSO
PMDA.
The structure
.B pmdaInterface
is used to convey the common procedural interface and state information that is
used by
.BR pmcd (1)
and a PMDA.
This state information includes tables describing the supported metrics and
instance domains.
.PP
As most of the
procedural interface is identical for all PMDAs,
they are provided as part of
this support library
.RB ( pmdaProfile (3),
.BR pmdaFetch (3),
.BR pmdaInstance (3),
.BR pmdaDesc (3),
.BR pmdaText (3)
and
.BR pmdaStore (3)).
However, these routines require access to the
.B pmdaInterface
state information so it must be correctly initialized using
.BR pmdaConnect (3),
.BR pmdaDaemon (3),
.BR pmdaOpenLog (3),
.BR pmdaDSO (3),
.BR pmdaGetOpt (3)
and
.BR pmdaInit (3).
.SH INSTANCES AND INSTANCE DOMAINS
Three structures are declared in
.I /usr/include/pcp/pmda.h
which provide a framework for declaring the metrics and instances supported by
the
PMDA.
.PP
Every instance requires a unique integer identifier and a unique name, as defined by
the structure
.BR pmdaInstid :
.PP
.nf
.ft CR
.in +0.5i
/*
 * Instance description: index and name
 */

typedef struct {
    int         i_inst;        /* internal instance identifier */
    char        *i_name;       /* external instance identifier */
} pmdaInstid;
.in
.fi
.PP
An instance domain requires its own unique identification
.RB ( pmInDom ),
the number of instances the domain represents, and a pointer to an array of
instance descriptions.  This is defined in the structure
.BR pmdaIndom :
.PP
.nf
.ft CR
.in +0.5i
/*
 * Instance domain description: unique instance id,
 * number of instances in this domain, and the list of
 * instances (not null terminated).
 */

typedef struct {
    pmInDom     it_indom;       /* indom, filled in */
    int         it_numinst;     /* number of instances */
    pmdaInstid  *it_set;        /* instance identifiers */
} pmdaIndom;
.in
.fi
.ft 1
.PP
The
.B simple PMDA
has one instance domain for
.IR simple . color
with three instances
.RI ( red ,
.I green
and
.IR blue ),
and a second instance domain for
.IR simple . now
with instances which can be specified at run-time.
These instance domains are defined as:
.PP
.nf
.ft CR
.in +0.5i
static pmdaInstid _color[] = {
    { 0, "red" }, { 1, "green" }, { 2, "blue" }
};
static pmdaInstid *_timenow = NULL;

static pmdaIndom indomtab[] = {
#define COLOR_INDOM	0
    { COLOR_INDOM, 3, _color },
#define NOW_INDOM 1
    { NOW_INDOM, 0, NULL },
};
.in
.fi
.PP
The preprocessor macros
.B COLOR_INDOM
and
.B NOW_INDOM
are used in the metric description table to identify the instance domains of
individual metrics.  These correspond to the
.I serial
value in the instance domain
.B pmInDom
structure (the
.I domain
field is set by
.BR pmdaInit (3)
at run-time).  The serial value must be unique for each instance domain
within the PMDA.
.PP
The indom table shown above which is usually passed to
.BR pmdaInit (3)
does not need to be created
if one wants to write one's own Fetch and Instance functions.
See
.BR pmdaInit (3)
for more details.
.SH NAMESPACE
Every PMDA has its own unique
.B namespace
using the format defined in
.BR PMNS (5).
In summary, the namespace matches the names of the metrics to the unique
identifier.  The
.B simple PMDA
defines five metrics:
.IR simple . numfetch ,
.IR simple . color ,
.IR simple . time . user,
.IR simple . time . sys
and
.IR simple . now .
The namespace for these metrics is defined in
.IR $PCP_PMDAS_DIR /simple/pmns
and is installed as:
.PP
.nf
.ft CR
.in +0.5in
simple {
    numfetch    253:0:0
    color       253:0:1
    time
    now         253:2:4
}

simple.time {
    user        253:1:2
    sys         253:1:3
}
.in
.fi
.PP
The domain number of
.I 253
is obtained from
.IR $PCP_VAR_DIR /pmns/stdpmid.
New PMDAs
should specify a unique domain number in this file, and obtain the
number during installation.  This allows the domain number to change by
modifying only the file
.IR $PCP_VAR_DIR /pmns/stdpmid.
.PP
The
.I simple.time
and
.I simple.now
metrics are defined in separate clusters to the other metrics which allows a
PMDA
to support more than 1024 metrics, as well as grouping similar metrics
together.  Therefore, the item numbers for a new cluster may be identical to
the item numbers in other clusters.  The
.B simple PMDA
continues to increment the item numbers to permit direct mapping (see
.BR pmdaInit (3)).
.PP
The namespace file should be installed and removed with the agent using
.BR pmnsadd (1)
and
.BR pmnsdel (1).
See the later sections on INSTALLATION and REMOVAL.
.PP
A simple ASCII namespace can be constructed by creating a file similar to
.IR $PCP_PMDAS_DIR /simple/root:
.PP
.nf
.ft CR
.in +0.5i
/*
 * fake "root" for validating the local PMNS subtree
 */

#include "$PCP_VAR_DIR/pmns/stdpmid"

root { simple }

#include "pmns"

.in
.fi
.PP
and can be referred to with the
.B \-n
option in most PCP tools.
.SH METRIC DESCRIPTIONS
Each metric requires a description
.RB ( pmDesc ),
which contains its
Performance Metric Identifier (PMID),
data type specification, instance domain, semantics
and units (see
.BR pmLookupDesc (3)).
A handle is also provided for application specific information in the
.B pmdaMetric
structure:
.PP
.nf
.ft CR
.in +0.5i
/*
 * Metric description: handle for extending description,
 * and the description.
 */

typedef struct {
    void*       m_user;         /* for users external use */
    pmDesc      m_desc;         /* metric description */
} pmdaMetric;
.in
.fi
.PP
The
.B simple PMDA
defines the metrics as:
.PP
.nf
.ft CR
.in +0.5i
static pmdaMetric metrictab[] = {
/* numfetch */
    { (void *)0,
      { PMDA_PMID(0,0), PM_TYPE_U32, PM_INDOM_NULL, PM_SEM_INSTANT,
        { 0,0,0,0,0,0} }, },
/* color */
    { (void *)0,
      { PMDA_PMID(0,1), PM_TYPE_32, COLOR_INDOM, PM_SEM_INSTANT,
        { 0,0,0,0,0,0} }, },
/* time.user */
    { (void*)0,
      { PMDA_PMID(1,2), PM_TYPE_DOUBLE, PM_INDOM_NULL, PM_SEM_COUNTER,
      	{ 0, 1, 0, 0, PM_TIME_SEC, 0 } }, },
/* time.sys */
    { (void*)0,
      { PMDA_PMID(1,3), PM_TYPE_DOUBLE, PM_INDOM_NULL, PM_SEM_COUNTER,
      	{ 0, 1, 0, 0, PM_TIME_SEC, 0 } }, },
/* now */
    { NULL,
      { PMDA_PMID(2,4), PM_TYPE_U32, NOW_INDOM, PM_SEM_INSTANT,
        { 0,0,0,0,0,0 } }, },
};
.in
.fi
.PP
The macro
.B PMDA_PMID
(defined in
.IR /usr/include/pcp/pmda.h )
is used to specify each metric's
.I cluster
and
.I item
fields of the associated
.BR pmID .
As with instance domains, the
.I domain
field is set by
.BR pmdaInit (3)
at run-time, however, the default domain is assumed to be defined by the
PMDA
in the macro
.BR MYDOMAIN .
.PP
The metric table shown above which is usually passed to
.BR pmdaInit (3)
does not need to be created
if one wants to write one's own Fetch and Descriptor functions.
See
.BR pmdaInit (3)
for more details.
.SH DSO PMDA
A
PMDA
that is run as a DSO is opened by
.BR pmcd (1)
with
.BR dlopen (3).
.BR pmcd (1)
will call the PMDA's
initialization function that is specified in
.IR $PCP_PMCDCONF_PATH .
This function is passed a pointer to a
.B pmdaInterface
structure which must be completed.  Any callbacks which are
.I not
the default
PMDA
support library callbacks must be specified in the
.B pmdaInterface
structure.
.PP
The
.B simple PMDA
uses its own store and fetch callback.
.BR simple_fetch ()
calls
.BR pmdaFetch (3)
which requires a callback to be set with
.BR pmdaSetFetchCallBack (3)
as can be seen in
.IR $PCP_PMDAS_DIR /simple/simple.c.
.PP
The flag
.B _isDSO
is used to determine if the
PMDA
is a daemon or a DSO so that the correct initialization
routine,
.BR pmdaDaemon (3)
or
.BR pmdaDSO (3),
is called.
.SH DAEMON PMDA
A
PMDA
that is run as a daemon is forked and executed by
.BR pmcd (1).
Therefore, unlike a DSO PMDA,
the starting point for a daemon PMDA is
.BR main ().
The agent should parse the command line arguments, create
a log file and initialize some data structures that
.BR pmcd (1)
would initialize for a DSO agent.
.PP
The
.B pmdaInterface
structure must be completely defined by the daemon PMDA.
The function
.BR pmdaDaemon (3)
can be called at the start of
.BR main ()
to set most of these fields.  Command line parsing can be simplified by using
.BR pmdaGetOpt (3),
which is similar to
.BR getopt (2),
but extracts a common set of options into the
.B pmdaInterface
structure.
.I stderr
can be mapped to a log file using
.BR pmdaOpenLog (3)
to simplify debugging and error messages.  The connection to
.BR pmcd (1)
can be made with
.BR pmdaConnect (3)
and the loop which handles the incoming PDUs,
.BR pmdaMain (3),
should be the last function called.  This can be seen in
.IR $PCP_PMDAS_DIR /simple/simple.c.
.PP
The
.BR simple_init ()
routine is common to an agent that can be run as both a Daemon and DSO
PMDA.
.SH HELP TEXT
Each PMDA must be able to provide
.BR pmcd (1)
with the help text for each metric.  Most PMDAs
use specially created files with indexes to support
efficient retrieval of the help text.
Tools are provided with PCP to
create the help text files of appropriate format. See
.BR newhelp (1).
.SH INSTALLATION AND REMOVAL
A series of shell procedures are defined in
.IR $PCP_SHARE_DIR /lib/pmdaproc.sh
which greatly simplify the installation and removal of a PMDA.
.PP
The
.I Install
scripts for most PMDAs
should only need to specify the name of the PMDA in
.BR iam ,
call
.B pmdaSetup
which sets up some default variables,
checks permissions (you have to be ``root'' to install or remove
a PMDA), checks that you're in the right directory (somewhere that ends
with /pmdas/\fI$iam\fP), optionally generate
the Performance Metrics Name Space (PMNS)
and PMDA domain
number files for Perl or Python PMDAs, checks the PMDA domain number is
valid, etc.,
specify the communication protocols,
and finally call
.B pmdaInstall
to do all the work of updating the PMNS, updating the
.BR pmcd (1)
control file, notifying or restarting
.BR pmcd (1),
.PP
Beyond
.B pmdaSetup
and
.BR pmdaInstall ,
another optional helper routines is
.B pmdaChooseConfigFile
that may be used to interactively select or create a PMDA-specific
configuration file,
e.g. \c
.BR pmdalogger (1).
.PP
The
.I Remove
scripts are even simpler as setting up the communication protocols
are not required, so set the name of the PMDA in
.IR iam ,
then call
.B pmdaSetup
followed by a call to
.BR pmdaRemove .
.PP
Further information is contained in the
.IR $PCP_SHARE_DIR /lib/pmdaproc.sh
file and the following section.
.PP
Optionally, a PMDA may provide an
.I Upgrade
script alongside
.I Install
and
.IR Remove .
If present this script will be used by the
.B pmcd
startup process to ensure corrections to an installation
have been made before starting the PMDA.
Examples of such corrections include: updates to
.I pmcd.conf
when a PMDA script or binary has been renamed, when the
PMDA supports a new format of its configuration file, or
if there is some latent problem from an earlier install
(e.g. some PMDAs may need to introduce use of the
.B notready
keyword in
.IR pmcd.conf ,
as described below).
.SH INSTALLATION REFINEMENTS
As outlined below
there are a number of variables that can be set in a PMDA's
.I Install
script to influence the behaviour of the installation procedures.
These would typically need to be set before the call to
.BR pmdaInstall ,
but in some instances (like
.I $iam
and the cases specifically noted below),
before the call to
.BR pmdaSetup .
.PP
The following variables control the communication options between the PMDA
and
.BR pmcd (1).
At least one of
.IR $daemon_opt ,
.IR $dso_opt ,
.I $perl_opt
or
.I $python_opt
must be set to define the supported mode(s) of communication.
If more than one of these is set the user will be prompted
to make a selection when the
.I Install
script is run.
.TP 16n
.I daemon_opt
We are willing to install the PMDA as a daemon.
.br
Default: true
.TP
.I dso_opt
We are willing to install the PMDA as a DSO, so
.BR pmcd (1)
will use
the dynamic linking loader to attach the PMDA's DSO at run-time and
communication from
.BR pmcd (1)
to the PMDA and back uses procedure calls, not an IPC channel.
.br
Default: false
.TP
.I dso_entry
For a DSO PMDA, this is the name of the PMDA's initialization routine.
.br
Default: \fI${iam}\fP_init
.TP
.I dso_name
For a DSO PMDA, this is the full pathanme of the PMDA's DSO file.
.br
Default: \fI$PCP_PMDAS_DIR\fP/\fI$iam\fP/pmda_\fI$iam\fP.\fI$dso_suffix\fP
.TP
.I pipe_opt
For a daemon PMDA, is the default IPC channel via a
.BR pipe (2)?
.br
Default: Platform-specific, so \fBtrue\fP for most, but \fBfalse\fP for Windows
.TP
.I perl_opt
We are willing to install the PMDA as a Perl script and
.BR pmcd (1)
will use the
.BR perl (1)
interpreter to run the PMDA.
.br
Default: false
.TP
.I pmda_dir
Full pathname to the directory where the PMDA's installation files
(executable, script, PMNS source, help text source, etc) are to
be found.
.br
Default: output from pwd(1)
.RS 16n
.PP
If set, must be done before the call to
.BR pmdaSetup .
.RE
.TP
.I pmda_name
For a daemon PMDA, this is the name of the PMDA's executable
binary relative to the
.I $pmda_dir
directory.
.br
Default: pmda\fI$iam\fP
.TP
.I python_opt
We are willing to install the PMDA as a Python script and
.BR pmcd (1)
will use the
.BR python (1)
interpreter to run the PMDA.
.br
Default: false
.TP
.I ipc_prot
For a daemon PMDA, this can be set to either
.B binary
or
.BR text .
The default is
.B binary
and
.B text
is rarely used.
In addition, an optional IPC parameter
.B notready
can be used to signify that the PMDA will start up in the
.B notready
state, e.g. \fBipc_prot="binary notready"\fP.
Note that the quotes are required.
The IPC parameters for a PMDA appear in
.B pmcd.conf
in the
.B "IPC Params"
column.
For further details, see
.BR pmcd (1)
but basically
.B pmcd
will not issue any requests to a PMDA that has started in the
.B notready
state until the PMDA sends a
.B PM_ERR_PMDAREADY
PDU.
This allows PMDAs with long startup times to initialize correctly
without timing out.
For details, see

CHANGELOG
pcp-7.0.3 (27 Nobember 2025)
    - Work-in-progress, not yet released.

pcp-7.0.2 (15 October 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: update metric removal mechanism
	pmdaopentelemetry: QA failure fixes and improvements
	pmdadocker: auto detect PCP_DOCKER_DIR using the Docker Engine API
	pmdahacluster: support "cibadmin --local" deprecation for pacemaker metrics
	pmdasmart: Better handling of non-supported NVME features
	pmdanvidia: add pmlogconf setups for logging GPU metrics
	pmdacifs: add support for collecting DebugData metrics
	pmdahdb: remove unused SPS version macros, various improvements
	pmdalinux: add new numastat per-node metrics (mem.numa.util.*)
	pmdarocestat: fix invalid path on RHEL8, support MLNX OFED 4.1
	pmdaproc: fix macOS ticks calculation for process time
	pmdaproc: updates to the macOS scheduler state metric
    - Client tools and utilities:
	pcp-meminfo: fix handling of hinv.hugepagesize which uses bytes
	pcp-atop: fix column width for CID when showing GPU stats (e)
	pcp-dstat: reduce width of GPU columns, more like the CPU columns
	pcp-dstat: fix incremental sample updating, add default time column
	pcp-xsos: source code merged from separate xsos-pcp repository
	pcp-htop: sync with latest htop.dev code fixes
	vllmbench2pcp: convert vLLM bench results to PCP archives
	guidellm2pcp: add run_id as a metric also, for ease of analysis
	guidellm2pcp: ensure instance identifier stability across runs
    - Server-side utilities and log management scripts:
	pmlogctl: fix variable renaming fallout
	pmlogconf: insert additional metrics needed for pcp-numastat
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix formatting arguments to xz_debug()
	libpcp3: fix formatting arguments to xz_debug()
	libpcp_web: eliminate possible use after free in series_calculate_rate()
	libpcp_web: conditional exports handling improvements
	libpcp_web: fix invalid HTTP request cache handling
	libpcp_web: fix partial HTTP response handling
	libpcp: remove unneeded conditional from __pmHashClear() [Coverity]
	pmieconf: eliminate minor memory leaks [Coverity]
    - Build, infrastructure, containers and packaging updates:
	build: update OpenIndiana platform support
	build: sync redhat.spec with recent rpm package changes
	build: packaging changes for openSUSE 16.0
	build: merge separate xsos-pcp repository into main PCP tree
	build: improvements for pcp-import-benchmarks packaging
	packaging: add transitional package pcp-import-guidellm2pcp
	packaging: FreeBSD and OpenBSD package list updates
    - Documentation:
	docs: extend pcp container readme to show nvidia GPU enablement
	docs: update INSTALL.md macOS pmcd start instructions
	docs: resolve warnings from man pages on macOS

pcp-7.0.1 (21 September 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: new per-process metrics agent for macOS
	pmdagfs2: Add support for glockfd debugfs metric collection
	pmdalinux: add NFSv4 connected clients and open files metrics
	pmdalinux: additional /proc/net/netstat MPTcp metrics
	pmdabpftrace: Adjust event loop code to work with Python 3.14
    - Client tools and utilities:
	guidellm2pcp: convert GuideLLM benchmark results to PCP archives
	pcp-mpstat: use vuser/vnice metrics and unify CPU utilization values
    - Server-side utilities and log management scripts:
	pmproxy, libpcp_web: ensure correct path used for conflict resolution
	pmlogger, pmie scripts: improvements for BSD date(1) differences
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix accounting within __pmHashCtl after walk-based-delete
	libpcp: fixes for >2GB archive files on 32-bit platforms
	python api: fix datetime-based timestamps in import API
	python api: improve exception handler in import API
    - Build, infrastructure, containers and packaging updates:
	build: automate macOS release and build process on github
	build: add aarch64 archive-analysis container builds on quay+ghcr
	build: debian conditional Build-Depends: for systemd pkgconfig file
	build: packaging list updates for RHEL 10 builds of PCP
	build: add pmproxy to the rpm set needing LOCAL_ONLY port setup
	build: ensure pcp-zeroconf rpm install script succeeds
	build: add zstd package as a hard rpm package dependency
	containers: archive-analysis container exports optional REST API
	containers: switch archive-analysis to using pcp-import-pmseries
	containers: add PCP_DOMAIN_AGENTS to pcp for auto-enabling PMDAs
	containers: make most/all PMDAs available in the pcp container
	containers: update the pmproxy template config in pcp container
    - Documentation:
	docs: fix openapi.yaml issues (pcp.readthedocs.io/en/latest/api)
	docs: add a containers note about alternate registry at ghcr.io
	docs: add a CLAUDE.md file for devs using claude code with PCP

pcp-7.0.0 (1 September 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaopentelemetry: new exporter for OpenTelemetry metrics
	pmdarocestat: new RDMA over Converged Ethernet (RoCE) metrics
	pmdahdb: new agent exporting SAP HANA metrics
	pmdalinux: additional TCP networking metrics
	pmdalinux: additional kernel memory and hugetlb metrics
	pmdalinux: support Amazon Linux within kernel.uname.distro
	pmdaopenmetrics: improvements to dynamic metric name handling
	pmdaopenmetrics: allow configuration files with epoch times
	pmdapodman: fix container name/command handling in new podman
	pmdahacluster: additional pacemaker node metrics
	pmdasmart: add smart.wwid.* metrics with drive UUID for indom
	pmdammv: support empty instance domains and dynamic map resets
	pmdammv: export instance domain help text when requested
    - Server-side utilities and log management scripts:
	pmlogger, pmproxy: support for a remote archive push webhook;
	  providing central logging without server-side configuration
	pmlogconf: updates to ensure complete sysstat metric coverage

src/cpan/PMDA.pm
package PCP::PMDA;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Exporter;
require DynaLoader;
require AutoLoader;

@ISA = qw( Exporter DynaLoader );
@EXPORT = qw(
    pmda_pmid pmda_pmid_name pmda_pmid_text pmda_units pmda_config pmda_uptime
    PM_ID_NULL PM_INDOM_NULL PM_IN_NULL
    PM_SPACE_BYTE PM_SPACE_KBYTE PM_SPACE_MBYTE PM_SPACE_GBYTE PM_SPACE_TBYTE
    PM_TIME_NSEC PM_TIME_USEC PM_TIME_MSEC PM_TIME_SEC PM_TIME_MIN PM_TIME_HOUR
    PM_COUNT_ONE
    PM_TYPE_NOSUPPORT PM_TYPE_32 PM_TYPE_U32 PM_TYPE_64 PM_TYPE_U64
	PM_TYPE_FLOAT PM_TYPE_DOUBLE PM_TYPE_STRING
    PM_SEM_COUNTER PM_SEM_INSTANT PM_SEM_DISCRETE
    PM_ERR_GENERIC PM_ERR_PMNS PM_ERR_NOPMNS PM_ERR_DUPPMNS PM_ERR_TEXT
	PM_ERR_APPVERSION PM_ERR_VALUE PM_ERR_LICENSE PM_ERR_TIMEOUT
	PM_ERR_NODATA PM_ERR_RESET PM_ERR_FILE PM_ERR_NAME PM_ERR_PMID
	PM_ERR_INDOM PM_ERR_INST PM_ERR_UNIT PM_ERR_CONV PM_ERR_TRUNC
	PM_ERR_SIGN PM_ERR_PROFILE PM_ERR_IPC PM_ERR_NOASCII PM_ERR_EOF
	PM_ERR_NOTHOST PM_ERR_EOL PM_ERR_MODE PM_ERR_LABEL PM_ERR_LOGREC
	PM_ERR_NOTARCHIVE PM_ERR_NOCONTEXT PM_ERR_PROFILESPEC PM_ERR_PMID_LOG
	PM_ERR_INDOM_LOG PM_ERR_INST_LOG PM_ERR_NOPROFILE PM_ERR_NOAGENT
	PM_ERR_PERMISSION PM_ERR_CONNLIMIT PM_ERR_AGAIN PM_ERR_ISCONN
	PM_ERR_NOTCONN PM_ERR_NEEDPORT PM_ERR_WANTACK PM_ERR_NONLEAF
	PM_ERR_PMDANOTREADY PM_ERR_PMDAREADY PM_ERR_OBJSTYLE PM_ERR_PMCDLICENSE
	PM_ERR_TOOSMALL PM_ERR_TOOBIG PM_ERR_NYI
);
@EXPORT_OK = qw( );
$VERSION = '1.00';

# metric identification
sub PM_ID_NULL		{ 0xffffffff; }
sub PM_INDOM_NULL	{ 0xffffffff; }
sub PM_IN_NULL		{ 0xffffffff; }

# units - space scale
sub PM_SPACE_BYTE	{ 0; }	# bytes
sub PM_SPACE_KBYTE	{ 1; }	# kilobytes
sub PM_SPACE_MBYTE	{ 2; }	# megabytes
sub PM_SPACE_GBYTE	{ 3; }	# gigabytes
sub PM_SPACE_TBYTE	{ 4; }	# terabytes

# units - time scale
sub PM_TIME_NSEC	{ 0; }	# nanoseconds
sub PM_TIME_USEC	{ 1; }	# microseconds
sub PM_TIME_MSEC	{ 2; }	# milliseconds
sub PM_TIME_SEC		{ 3; }	# seconds
sub PM_TIME_MIN		{ 4; }	# minutes
sub PM_TIME_HOUR	{ 5; }	# hours

# units - count scale	(for metrics such as count events, syscalls,
# interrupts, etc - these are simply powers of ten and not enumerated here
# (e.g. 6 for 10^6, or -3 for 10^-3).
sub PM_COUNT_ONE	{ 0; }	# 1

# data type of metric values
sub PM_TYPE_NOSUPPORT	{ 0xffffffff; }	# not implemented in this version
sub PM_TYPE_32		{ 0; }	# 32-bit signed integer
sub PM_TYPE_U32		{ 1; }	# 32-bit unsigned integer
sub PM_TYPE_64		{ 2; }	# 64-bit signed integer
sub PM_TYPE_U64		{ 3; }	# 64-bit unsigned integer
sub PM_TYPE_FLOAT	{ 4; }	# 32-bit floating point
sub PM_TYPE_DOUBLE	{ 5; }	# 64-bit floating point
sub PM_TYPE_STRING	{ 6; }	# array of characters

# semantics/interpretation of metric values
sub PM_SEM_COUNTER	{ 1; }	# cumulative counter (monotonic increasing)
sub PM_SEM_INSTANT	{ 3; }	# instantaneous value, continuous domain
sub PM_SEM_DISCRETE	{ 4; }	# instantaneous value, discrete domain

# error codes
sub PM_ERR_GENERIC	{ -12345; }	# Generic error, already reported above
sub PM_ERR_PMNS		{ -12346; }	# Problems parsing PMNS definitions
sub PM_ERR_NOPMNS	{ -12347; }	# PMNS not accessible
sub PM_ERR_DUPPMNS	{ -12348; }	# Attempt to reload the PMNS
sub PM_ERR_TEXT		{ -12349; }	# Oneline or help text is not available
sub PM_ERR_APPVERSION	{ -12350; }	# Metric not supported by this version of monitored application
sub PM_ERR_VALUE	{ -12351; }	# Missing metric value(s)
sub PM_ERR_LICENSE	{ -12352; }	# Current PCP license does not permit this operation
sub PM_ERR_TIMEOUT	{ -12353; }	# Timeout waiting for a response from PMCD
sub PM_ERR_NODATA	{ -12354; }	# Empty archive log file
sub PM_ERR_RESET	{ -12355; }	# PMCD reset or configuration change
sub PM_ERR_FILE		{ -12356; }	# Cannot locate a file
sub PM_ERR_NAME		{ -12357; }	# Unknown metric name
sub PM_ERR_PMID		{ -12358; }	# Unknown or illegal metric identifier
sub PM_ERR_INDOM	{ -12359; }	# Unknown or illegal instance domain identifier
sub PM_ERR_INST		{ -12360; }	# Unknown or illegal instance identifier
sub PM_ERR_UNIT		{ -12361; }	# Illegal pmUnits specification
sub PM_ERR_CONV		{ -12362; }	# Impossible value or scale conversion
sub PM_ERR_TRUNC	{ -12363; }	# Truncation in value conversion
sub PM_ERR_SIGN		{ -12364; }	# Negative value in conversion to unsigned
sub PM_ERR_PROFILE	{ -12365; }	# Explicit instance identifier(s) required
sub PM_ERR_IPC		{ -12366; }	# IPC protocol failure
sub PM_ERR_NOASCII	{ -12367; }	# ASCII format not supported for this PDU
sub PM_ERR_EOF		{ -12368; }	# IPC channel closed
sub PM_ERR_NOTHOST	{ -12369; }	# Operation requires context with host source of metrics
sub PM_ERR_EOL		{ -12370; }	# End of PCP archive log
sub PM_ERR_MODE		{ -12371; }	# Illegal mode specification
sub PM_ERR_LABEL	{ -12372; }	# Illegal label record at start of a PCP archive log file
sub PM_ERR_LOGREC	{ -12373; }	# Corrupted record in a PCP archive log
sub PM_ERR_NOTARCHIVE	{ -12374; }	# Operation requires context with archive source of metrics
sub PM_ERR_NOCONTEXT	{ -12376; }	# Attempt to use an illegal context
sub PM_ERR_PROFILESPEC	{ -12377; }	# NULL pmInDom with non-NULL instlist
sub PM_ERR_PMID_LOG	{ -12378; }	# Metric not defined in the PCP archive log
sub PM_ERR_INDOM_LOG	{ -12379; }	# Instance domain identifier not defined in the PCP archive log
sub PM_ERR_INST_LOG	{ -12380; }	# Instance identifier not defined in the PCP archive log
sub PM_ERR_NOPROFILE	{ -12381; }	# Missing profile - protocol botch
sub PM_ERR_NOAGENT	{ -12386; }	# No PMCD agent for domain of request
sub PM_ERR_PERMISSION	{ -12387; }	# No permission to perform requested operation

sub PM_ERR_CONNLIMIT	{ -12388; }	# PMCD connection limit for this host exceeded
sub PM_ERR_AGAIN	{ -12389; }	# Try again. Information not currently available
sub PM_ERR_ISCONN	{ -12390; }	# Already Connected
sub PM_ERR_NOTCONN	{ -12391; }	# Not Connected
sub PM_ERR_NEEDPORT	{ -12392; }	# A non-null port name is required
sub PM_ERR_WANTACK	{ -12393; }	# Cannot send due to pending acknowledgements
sub PM_ERR_NONLEAF	{ -12394; }	# Metric name is not a leaf in PMNS
sub PM_ERR_PMDANOTREADY	{ -13394; }	# PMDA is not yet ready to respond to requests
sub PM_ERR_PMDAREADY	{ -13393; }	# PMDA is now responsive to requests
sub PM_ERR_OBJSTYLE	{ -12395; }	# Caller does not match object style of running kernel
sub PM_ERR_PMCDLICENSE	{ -12396; }	# PMCD is not licensed to accept client connections
sub PM_ERR_TOOSMALL	{ -12443; }	# Insufficient elements in list
sub PM_ERR_TOOBIG	{ -12444; }	# Result size exceeded
sub PM_ERR_NYI		{ -21344; }	# Functionality not yet implemented


bootstrap PCP::PMDA $VERSION;

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

PCP::PMDA - Perl extension for Performance Metrics Domain Agents

=head1 SYNOPSIS

  use PCP::PMDA;
  blah blah blah

=head1 DESCRIPTION







=head1 AUTHOR

nathans@debian.org

=head1 SEE ALSO

perl(1) and pmda(3).

=cut
